### ISSUE TITLE
Auto add to permanent ignore (d/e fails) false positives

### ISSUE THREAD

**Issue by Road-block** [osilvay/DisenchanterPlus#5]
There are no Lua errors so this is more of a design/logic error.

I figure the `UNIT_SPELLCAST_FAILED` handler has been added to try and catch items that cannot be disenchanted (like points vendor items, some pvp items etc)

The problem is there's at least 2 ways that a disenchant cast will commonly fail that has nothing to do with the item attempted.

1. On a flying mount with auto-dismount disabled. DisenchantPlus window pops up, hit Proceed, cast fails item gets added to permanent ignore.
2. Autorun + hit proceed. Cast fails because you cannot d/e while moving. Item gets added to permalist.

There's likely other situations (taxi?) where the d/e cast can fail but the item is still perfectly disenchantable.

**Comment by Road-block** [comment]
This is a tricky problem to solve.

My best guess would be you'd need to monitor at least 2 sets of events for this scenario.
The `UNIT_SPELLCAST_FAILED` event should __only__ add an item to the ignore list if the castGUID parameter can be matched to a preceding `UNIT_SPELLCAST_SENT` event for disenchant spellid with the __same castGUID__

This is the only situation where a server roundtrip is required to get the "can't disenchant this" information.
Cast gets sent > server replies "you cannot disenchant this" and the cast fails.

For other situations (mounted / moving / taxi / pacified / dead etc) the client can decide on its own that this action is not available and the cast is __never sent__.

TL;DR: Only `UNIT_SPELLCAST_FAILED` events with a matching `UNIT_SPELLCAST_SENT` stored previously with the same castGUID to detect items not able to be disenchanted. `UNIT_SPELLCAST_FAILED` that do not follow a `UNIT_SPELLCAST_SENT` should be ignored.

**Comment by Road-block** [comment]
Minimal implementation.
```lua
local disenchantCastGUID = nil
```
added under [DisenchantProcess.lua#L23](https://github.com/osilvay/DisenchanterPlus/blob/97fe4cbecb8579240d4cd9eac32259d1637bd6a2/Modules/Disenchant/DisenchantProcess.lua#L23)
where the other variables are initialized.
Then :UnitSpellCastSent at [DisenchantProcess.lua#L101](https://github.com/osilvay/DisenchanterPlus/blob/97fe4cbecb8579240d4cd9eac32259d1637bd6a2/Modules/Disenchant/DisenchantProcess.lua#L101) changed to
```lua
function DP_DisenchantProcess:UnitSpellCastSent(unit, target, castGUID, spellID)
  --DisenchanterPlus:Debug(string.format("UnitSpellCastStart : unitTarget = %s, castGUID = %s, spellID = %d", unitTarget, castGUID, spellID))
  if unit == "player" and spellID == disenchantSpellID then
    disenchantCastGUID = castGUID
  end
end
```
and finally [DisenchantProcess.lua#L189](https://github.com/osilvay/DisenchanterPlus/blob/97fe4cbecb8579240d4cd9eac32259d1637bd6a2/Modules/Disenchant/DisenchantProcess.lua#L189) changed to
```lua
function DP_DisenchantProcess:UnitSpellCastFailed(unitTarget, castGUID, spellID)
  if unitTarget == "player" and spellID == disenchantSpellID then
    if castGUID and (castGUID ~= disenchantCastGUID) then return end -- < this is the addition
    if not disenchanting then disenchanting = true end
```

Tested by 
- attempting to disenchant rep vendor items (that cannot be disenchanted) they properly add to ignore list.
- attempting to disenchant an uncommon item (that can be disenchanted) while mounted / moving etc. Error text indicates why the disenchant failed and item is not added to permanent ignore (correct behaviour)

**Comment by osilvay** [comment]
Thanks for the report. I've uploaded version 1.0.9.0 with these changes you mention to github (and with other code changes), including a massive enchantment window.
I haven't been able to test this last one in cataclysm... but it probably fails because of the way the enchantment lists are set up between cataclysm and classic era.
I'm still working on this, that's why I haven't released a version yet.
